extensions/amp-bind/amp-bind.md

```diff
@@ -218,32 +278,143 @@ Only binding to the following components and attributes are allowed. Most bindab
+<p [class]="'abc'" class="def"></p>
+```
+
+In development mode, `amp-bind` will also issue a warning when dereferencing undefined variables or properties. This can also help prevent unintended mutations due to `null` expression results. For example:
+
+```html
+<amp-state id="myAmpState">
+  <script type="application/json">
+    { "foo": 123 }
+  </script>
+</amp-state>
+
+<!-- The amp-state#myAmpState does not have a `bar` variable, so a warning
+     will be issued in development mode. -->
+<p [text]="myAmpState.bar">Some placeholder text.</p>
+```
+
+### Errors
+
+There are several types of runtime errors that may be encountered when working with `amp-bind`.
+
+| Type | Example | Message | Suggestion |
+| --- | --- | --- | --- |
+| Invalid binding | `<p [someBogusAttribute]="myExpression">` | *Binding to [someBogusAttribute] on `<P>`` is not allowed.* | Make sure that only [whitelisted bindings](#element-specific-attributes) are used. |
+| Syntax error | `<p [text]="(missingClosingParens">` | *Expression compilation error in...* | Double-check the expression for typos. |
+| Non-whitelisted functions | `<p [text]="alert(1)"></p>` | *alert is not a supported function.* | Only use [whitelisted functions](#whitelisted-functions). |
+| Sanitized result | `<a href="javascript:alert(1)"></a>` | *"javascript:alert(1)" is not a valid result for [href].* | Avoid banned URL protocols or expressions that would fail the AMP Validator. |
+
+## Appendix
+
+### Builtin Functions
+
+`amp-bind` includes the following builtin functions.
+
+| Name | Description | Arguments | Examples |
+| --- | --- | --- | --- |
+| `copyAndSplice` | Similar to [Array#splice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) except a copy of the spliced array is returned. | `array` : An array.<br>`start` : Index at which to start changing the array.<br> `deleteCount` : The number of items to delete, starting at index `start`<br> `items...` Items to add to the array, beginning at index `start`. | `// Deleting an element. Returns [1, 3]`<br>`copyAndSplice([1, 2, 3], 1, 1)`<br>`// Replacing an item. Returns ['Pizza', 'Cake', 'Ice Cream']`<br>`copyAndSplice(['Pizza', 'Cake', 'Soda'], 2, 1, 'Ice Cream')` |
+
+### Deep State Merging
+
+`amp-bind` deep-merges all state updates with bind's current state. Whenevr `amp-bind` receives a state update, all variables from the update are written to `amp-bind`'s state with one important exception. If `amp-bind`'s state and the state update both contain the same variable **AND** both variable definitions are objects, then the objects are recursively merged to a maximum depth of ten.
+
+Consider the following example:
+
+```javascript
+{
+<!-- State is empty -->
+}
+```
+
+```html
+<button on="tap:AMP.setState({employee: {name: 'John Smith', address: '123 Fake Street', age: 47, vehicle: 'Car'}})"...></button>
+<button on="tap:AMP.setState({employee: {age: 48}})"...></button>
```

Comment:
Minor: Remove "vehicle" and "address" keys and make the new age very different from the first.